pages/en/lb4/Dependency-injection.md

```diff
@@ -140,28 +140,101 @@ class InfoController {
 }
 ```
 
+## Optional dependencies
+
+Sometimes the dependencies are optional. For example, the logging level for a Logger provider can have a default value if it is not set (bound to the context). 
+
+To resolve an optional dependency, set `optional` flag to true:
+
+```ts
+const ctx = new Context();
+await ctx.get('optional-key', {optional: true}); // Return `undefined` instead of throwing an error
+```
+
+Here is anthoer example with dependency injection:
+
+```ts
+export class LoggerProvider implements Provider<Logger> {
+  // Log writer is an optional dependency and it falls back to `logToConsole`
+  @inject('log.writer', {optional: true})
+  private logWriter: LogWriterFn = logToConsole;
+
+  @inject('log.level', {optional: true})
+  private logLevel: string = 'WARN';
+}
+```
+
+Optional dependencies can be used with constructor, property, and method injections. You can use default values with such cases. For example:
+
+```ts
+export class LoggerProvider implements Provider<Logger> {
+  constructor(
+    // Log writer is an optional dependency and it falls back to `logToConsole`
+    @inject('log.writer', {optional: true})
+    private logWriter: LogWriterFn = logToConsole,
+
+    @inject('log.level', {optional: true})
+    private logLevel: string = 'WARN',
```

Comment:
AFAICT, this code example is exactly the same as the code example above. Is this intentional?

Did you perhaps want to show constructor-level and method-level injections as an addition to property-based injection shown previously?

Reply:
Yes.

Reply:
I am confused. The sentence before this code block says

> Optional dependencies can also be used with constructor and method injections. For example:

But the code example shows property-level injections only.

Could you please clarify?

Reply:
You probably didn't notice the 2nd example has `constructor(...)`.